// @seamapi/http/src/lib/seam/connect/routes/acs-users.ts
/*
 * Automatically generated by generate-routes.ts.
 * Do not edit this file or add other files to this directory.
 */

import type { RouteRequestBody, RouteResponse } from '@seamapi/types/connect'
import type { SetNonNullable } from 'type-fest'

import {
  getAuthHeadersForClientSessionToken,
  warnOnInsecureuserIdentifierKey,
} from 'lib/seam/connect/auth.js'
import { type Client, createClient } from 'lib/seam/connect/client.js'
import {
  isSeamHttpOptionsWithApiKey,
  isSeamHttpOptionsWithClient,
  isSeamHttpOptionsWithClientSessionToken,
  isSeamHttpOptionsWithConsoleSessionToken,
  isSeamHttpOptionsWithPersonalAccessToken,
  type SeamHttpFromPublishableKeyOptions,
  SeamHttpInvalidOptionsError,
  type SeamHttpOptions,
  type SeamHttpOptionsWithApiKey,
  type SeamHttpOptionsWithClient,
  type SeamHttpOptionsWithClientSessionToken,
  type SeamHttpOptionsWithConsoleSessionToken,
  type SeamHttpOptionsWithPersonalAccessToken,
  type SeamHttpRequestOptions,
} from 'lib/seam/connect/options.js'
import {
  limitToSeamHttpRequestOptions,
  parseOptions,
} from 'lib/seam/connect/parse-options.js'
import { SeamHttpRequest } from 'lib/seam/connect/seam-http-request.js'

import { SeamHttpClientSessions } from './client-sessions.js'

export class SeamHttpAcsUsers {
  client: Client
  readonly defaults: Required<SeamHttpRequestOptions>

  constructor(apiKeyOrOptions: string | SeamHttpOptions = {}) {
    const options = parseOptions(apiKeyOrOptions)
    this.client = 'client' in options ? options.client : createClient(options)
    this.defaults = limitToSeamHttpRequestOptions(options)
  }

  static fromClient(
    client: SeamHttpOptionsWithClient['client'],
    options: Omit<SeamHttpOptionsWithClient, 'client'> = {},
  ): SeamHttpAcsUsers {
    const constructorOptions = { ...options, client }
    if (!isSeamHttpOptionsWithClient(constructorOptions)) {
      throw new SeamHttpInvalidOptionsError('Missing client')
    }
    return new SeamHttpAcsUsers(constructorOptions)
  }

  static fromApiKey(
    apiKey: SeamHttpOptionsWithApiKey['apiKey'],
    options: Omit<SeamHttpOptionsWithApiKey, 'apiKey'> = {},
  ): SeamHttpAcsUsers {
    const constructorOptions = { ...options, apiKey }
    if (!isSeamHttpOptionsWithApiKey(constructorOptions)) {
      throw new SeamHttpInvalidOptionsError('Missing apiKey')
    }
    return new SeamHttpAcsUsers(constructorOptions)
  }

  static fromClientSessionToken(
    clientSessionToken: SeamHttpOptionsWithClientSessionToken['clientSessionToken'],
    options: Omit<
      SeamHttpOptionsWithClientSessionToken,
      'clientSessionToken'
    > = {},
  ): SeamHttpAcsUsers {
    const constructorOptions = { ...options, clientSessionToken }
    if (!isSeamHttpOptionsWithClientSessionToken(constructorOptions)) {
      throw new SeamHttpInvalidOptionsError('Missing clientSessionToken')
    }
    return new SeamHttpAcsUsers(constructorOptions)
  }

  static async fromPublishableKey(
    publishableKey: string,
    userIdentifierKey: string,
    options: SeamHttpFromPublishableKeyOptions = {},
  ): Promise<SeamHttpAcsUsers> {
    warnOnInsecureuserIdentifierKey(userIdentifierKey)
    const clientOptions = parseOptions({ ...options, publishableKey })
    if (isSeamHttpOptionsWithClient(clientOptions)) {
      throw new SeamHttpInvalidOptionsError(
        'The client option cannot be used with SeamHttp.fromPublishableKey',
      )
    }
    const client = createClient(clientOptions)
    const clientSessions = SeamHttpClientSessions.fromClient(client)
    const { token } = await clientSessions.getOrCreate({
      user_identifier_key: userIdentifierKey,
    })
    return SeamHttpAcsUsers.fromClientSessionToken(token, options)
  }

  static fromConsoleSessionToken(
    consoleSessionToken: SeamHttpOptionsWithConsoleSessionToken['consoleSessionToken'],
    workspaceId: SeamHttpOptionsWithConsoleSessionToken['workspaceId'],
    options: Omit<
      SeamHttpOptionsWithConsoleSessionToken,
      'consoleSessionToken' | 'workspaceId'
    > = {},
  ): SeamHttpAcsUsers {
    const constructorOptions = { ...options, consoleSessionToken, workspaceId }
    if (!isSeamHttpOptionsWithConsoleSessionToken(constructorOptions)) {
      throw new SeamHttpInvalidOptionsError(
        'Missing consoleSessionToken or workspaceId',
      )
    }
    return new SeamHttpAcsUsers(constructorOptions)
  }

  static fromPersonalAccessToken(
    personalAccessToken: SeamHttpOptionsWithPersonalAccessToken['personalAccessToken'],
    workspaceId: SeamHttpOptionsWithPersonalAccessToken['workspaceId'],
    options: Omit<
      SeamHttpOptionsWithPersonalAccessToken,
      'personalAccessToken' | 'workspaceId'
    > = {},
  ): SeamHttpAcsUsers {
    const constructorOptions = { ...options, personalAccessToken, workspaceId }
    if (!isSeamHttpOptionsWithPersonalAccessToken(constructorOptions)) {
      throw new SeamHttpInvalidOptionsError(
        'Missing personalAccessToken or workspaceId',
      )
    }
    return new SeamHttpAcsUsers(constructorOptions)
  }

  async updateClientSessionToken(
    clientSessionToken: SeamHttpOptionsWithClientSessionToken['clientSessionToken'],
  ): Promise<void> {
    const { headers } = this.client.defaults
    const authHeaders = getAuthHeadersForClientSessionToken({
      clientSessionToken,
    })
    for (const key of Object.keys(authHeaders)) {
      if (headers[key] == null) {
        throw new Error(
          'Cannot update a clientSessionToken on a client created without a clientSessionToken',
        )
      }
    }
    this.client.defaults.headers = { ...headers, ...authHeaders }
    const clientSessions = SeamHttpClientSessions.fromClient(this.client)
    await clientSessions.get()
  }

  addToAccessGroup(
    body?: AcsUsersAddToAccessGroupBody,
  ): SeamHttpRequest<void, undefined> {
    return new SeamHttpRequest(this, {
      path: '/acs/users/add_to_access_group',
      method: 'post',
      body,
      responseKey: undefined,
    })
  }

  create(
    body?: AcsUsersCreateBody,
  ): SeamHttpRequest<AcsUsersCreateResponse, 'acs_user'> {
    return new SeamHttpRequest(this, {
      path: '/acs/users/create',
      method: 'post',
      body,
      responseKey: 'acs_user',
    })
  }

  delete(body?: AcsUsersDeleteBody): SeamHttpRequest<void, undefined> {
    return new SeamHttpRequest(this, {
      path: '/acs/users/delete',
      method: 'post',
      body,
      responseKey: undefined,
    })
  }

  get(
    body?: AcsUsersGetParams,
  ): SeamHttpRequest<AcsUsersGetResponse, 'acs_user'> {
    return new SeamHttpRequest(this, {
      path: '/acs/users/get',
      method: 'post',
      body,
      responseKey: 'acs_user',
    })
  }

  list(
    body?: AcsUsersListParams,
  ): SeamHttpRequest<AcsUsersListResponse, 'acs_users'> {
    return new SeamHttpRequest(this, {
      path: '/acs/users/list',
      method: 'post',
      body,
      responseKey: 'acs_users',
    })
  }

  listAccessibleEntrances(
    body?: AcsUsersListAccessibleEntrancesParams,
  ): SeamHttpRequest<AcsUsersListAccessibleEntrancesResponse, 'acs_entrances'> {
    return new SeamHttpRequest(this, {
      path: '/acs/users/list_accessible_entrances',
      method: 'post',
      body,
      responseKey: 'acs_entrances',
    })
  }

  removeFromAccessGroup(
    body?: AcsUsersRemoveFromAccessGroupBody,
  ): SeamHttpRequest<void, undefined> {
    return new SeamHttpRequest(this, {
      path: '/acs/users/remove_from_access_group',
      method: 'post',
      body,
      responseKey: undefined,
    })
  }

  revokeAccessToAllEntrances(
    body?: AcsUsersRevokeAccessToAllEntrancesBody,
  ): SeamHttpRequest<void, undefined> {
    return new SeamHttpRequest(this, {
      path: '/acs/users/revoke_access_to_all_entrances',
      method: 'post',
      body,
      responseKey: undefined,
    })
  }

  suspend(body?: AcsUsersSuspendBody): SeamHttpRequest<void, undefined> {
    return new SeamHttpRequest(this, {
      path: '/acs/users/suspend',
      method: 'post',
      body,
      responseKey: undefined,
    })
  }

  unsuspend(body?: AcsUsersUnsuspendBody): SeamHttpRequest<void, undefined> {
    return new SeamHttpRequest(this, {
      path: '/acs/users/unsuspend',
      method: 'post',
      body,
      responseKey: undefined,
    })
  }

  update(body?: AcsUsersUpdateBody): SeamHttpRequest<void, undefined> {
    return new SeamHttpRequest(this, {
      path: '/acs/users/update',
      method: 'post',
      body,
      responseKey: undefined,
    })
  }
}

export type AcsUsersAddToAccessGroupBody =
  RouteRequestBody<'/acs/users/add_to_access_group'>

export type AcsUsersAddToAccessGroupResponse = SetNonNullable<
  Required<RouteResponse<'/acs/users/add_to_access_group'>>
>

export type AcsUsersAddToAccessGroupOptions = never

export type AcsUsersCreateBody = RouteRequestBody<'/acs/users/create'>

export type AcsUsersCreateResponse = SetNonNullable<
  Required<RouteResponse<'/acs/users/create'>>
>

export type AcsUsersCreateOptions = never

export type AcsUsersDeleteBody = RouteRequestBody<'/acs/users/delete'>

export type AcsUsersDeleteResponse = SetNonNullable<
  Required<RouteResponse<'/acs/users/delete'>>
>

export type AcsUsersDeleteOptions = never

export type AcsUsersGetParams = RouteRequestBody<'/acs/users/get'>

export type AcsUsersGetResponse = SetNonNullable<
  Required<RouteResponse<'/acs/users/get'>>
>

export type AcsUsersGetOptions = never

export type AcsUsersListParams = RouteRequestBody<'/acs/users/list'>

export type AcsUsersListResponse = SetNonNullable<
  Required<RouteResponse<'/acs/users/list'>>
>

export type AcsUsersListOptions = never

export type AcsUsersListAccessibleEntrancesParams =
  RouteRequestBody<'/acs/users/list_accessible_entrances'>

export type AcsUsersListAccessibleEntrancesResponse = SetNonNullable<
  Required<RouteResponse<'/acs/users/list_accessible_entrances'>>
>

export type AcsUsersListAccessibleEntrancesOptions = never

export type AcsUsersRemoveFromAccessGroupBody =
  RouteRequestBody<'/acs/users/remove_from_access_group'>

export type AcsUsersRemoveFromAccessGroupResponse = SetNonNullable<
  Required<RouteResponse<'/acs/users/remove_from_access_group'>>
>

export type AcsUsersRemoveFromAccessGroupOptions = never

export type AcsUsersRevokeAccessToAllEntrancesBody =
  RouteRequestBody<'/acs/users/revoke_access_to_all_entrances'>

export type AcsUsersRevokeAccessToAllEntrancesResponse = SetNonNullable<
  Required<RouteResponse<'/acs/users/revoke_access_to_all_entrances'>>
>

export type AcsUsersRevokeAccessToAllEntrancesOptions = never

export type AcsUsersSuspendBody = RouteRequestBody<'/acs/users/suspend'>

export type AcsUsersSuspendResponse = SetNonNullable<
  Required<RouteResponse<'/acs/users/suspend'>>
>

export type AcsUsersSuspendOptions = never

export type AcsUsersUnsuspendBody = RouteRequestBody<'/acs/users/unsuspend'>

export type AcsUsersUnsuspendResponse = SetNonNullable<
  Required<RouteResponse<'/acs/users/unsuspend'>>
>

export type AcsUsersUnsuspendOptions = never

export type AcsUsersUpdateBody = RouteRequestBody<'/acs/users/update'>

export type AcsUsersUpdateResponse = SetNonNullable<
  Required<RouteResponse<'/acs/users/update'>>
>

export type AcsUsersUpdateOptions = never
