// @seamapi/http/src/lib/seam/connect/routes/connect-webviews.ts
/*
 * Automatically generated by generate-routes.ts.
 * Do not edit this file or add other files to this directory.
 */

import type {
  RouteRequestBody,
  RouteRequestParams,
  RouteResponse,
} from '@seamapi/types/connect'
import type { SetNonNullable } from 'type-fest'

import {
  getAuthHeadersForClientSessionToken,
  warnOnInsecureuserIdentifierKey,
} from 'lib/seam/connect/auth.js'
import { type Client, createClient } from 'lib/seam/connect/client.js'
import {
  isSeamHttpOptionsWithApiKey,
  isSeamHttpOptionsWithClient,
  isSeamHttpOptionsWithClientSessionToken,
  isSeamHttpOptionsWithConsoleSessionToken,
  isSeamHttpOptionsWithPersonalAccessToken,
  type SeamHttpFromPublishableKeyOptions,
  SeamHttpInvalidOptionsError,
  type SeamHttpOptions,
  type SeamHttpOptionsWithApiKey,
  type SeamHttpOptionsWithClient,
  type SeamHttpOptionsWithClientSessionToken,
  type SeamHttpOptionsWithConsoleSessionToken,
  type SeamHttpOptionsWithPersonalAccessToken,
  type SeamHttpRequestOptions,
} from 'lib/seam/connect/options.js'
import {
  limitToSeamHttpRequestOptions,
  parseOptions,
} from 'lib/seam/connect/parse-options.js'
import { SeamHttpRequest } from 'lib/seam/connect/seam-http-request.js'

import { SeamHttpClientSessions } from './client-sessions.js'

export class SeamHttpConnectWebviews {
  client: Client
  readonly defaults: Required<SeamHttpRequestOptions>

  constructor(apiKeyOrOptions: string | SeamHttpOptions = {}) {
    const options = parseOptions(apiKeyOrOptions)
    this.client = 'client' in options ? options.client : createClient(options)
    this.defaults = limitToSeamHttpRequestOptions(options)
  }

  static fromClient(
    client: SeamHttpOptionsWithClient['client'],
    options: Omit<SeamHttpOptionsWithClient, 'client'> = {},
  ): SeamHttpConnectWebviews {
    const constructorOptions = { ...options, client }
    if (!isSeamHttpOptionsWithClient(constructorOptions)) {
      throw new SeamHttpInvalidOptionsError('Missing client')
    }
    return new SeamHttpConnectWebviews(constructorOptions)
  }

  static fromApiKey(
    apiKey: SeamHttpOptionsWithApiKey['apiKey'],
    options: Omit<SeamHttpOptionsWithApiKey, 'apiKey'> = {},
  ): SeamHttpConnectWebviews {
    const constructorOptions = { ...options, apiKey }
    if (!isSeamHttpOptionsWithApiKey(constructorOptions)) {
      throw new SeamHttpInvalidOptionsError('Missing apiKey')
    }
    return new SeamHttpConnectWebviews(constructorOptions)
  }

  static fromClientSessionToken(
    clientSessionToken: SeamHttpOptionsWithClientSessionToken['clientSessionToken'],
    options: Omit<
      SeamHttpOptionsWithClientSessionToken,
      'clientSessionToken'
    > = {},
  ): SeamHttpConnectWebviews {
    const constructorOptions = { ...options, clientSessionToken }
    if (!isSeamHttpOptionsWithClientSessionToken(constructorOptions)) {
      throw new SeamHttpInvalidOptionsError('Missing clientSessionToken')
    }
    return new SeamHttpConnectWebviews(constructorOptions)
  }

  static async fromPublishableKey(
    publishableKey: string,
    userIdentifierKey: string,
    options: SeamHttpFromPublishableKeyOptions = {},
  ): Promise<SeamHttpConnectWebviews> {
    warnOnInsecureuserIdentifierKey(userIdentifierKey)
    const clientOptions = parseOptions({ ...options, publishableKey })
    if (isSeamHttpOptionsWithClient(clientOptions)) {
      throw new SeamHttpInvalidOptionsError(
        'The client option cannot be used with SeamHttp.fromPublishableKey',
      )
    }
    const client = createClient(clientOptions)
    const clientSessions = SeamHttpClientSessions.fromClient(client)
    const { token } = await clientSessions.getOrCreate({
      user_identifier_key: userIdentifierKey,
    })
    return SeamHttpConnectWebviews.fromClientSessionToken(token, options)
  }

  static fromConsoleSessionToken(
    consoleSessionToken: SeamHttpOptionsWithConsoleSessionToken['consoleSessionToken'],
    workspaceId: SeamHttpOptionsWithConsoleSessionToken['workspaceId'],
    options: Omit<
      SeamHttpOptionsWithConsoleSessionToken,
      'consoleSessionToken' | 'workspaceId'
    > = {},
  ): SeamHttpConnectWebviews {
    const constructorOptions = { ...options, consoleSessionToken, workspaceId }
    if (!isSeamHttpOptionsWithConsoleSessionToken(constructorOptions)) {
      throw new SeamHttpInvalidOptionsError(
        'Missing consoleSessionToken or workspaceId',
      )
    }
    return new SeamHttpConnectWebviews(constructorOptions)
  }

  static fromPersonalAccessToken(
    personalAccessToken: SeamHttpOptionsWithPersonalAccessToken['personalAccessToken'],
    workspaceId: SeamHttpOptionsWithPersonalAccessToken['workspaceId'],
    options: Omit<
      SeamHttpOptionsWithPersonalAccessToken,
      'personalAccessToken' | 'workspaceId'
    > = {},
  ): SeamHttpConnectWebviews {
    const constructorOptions = { ...options, personalAccessToken, workspaceId }
    if (!isSeamHttpOptionsWithPersonalAccessToken(constructorOptions)) {
      throw new SeamHttpInvalidOptionsError(
        'Missing personalAccessToken or workspaceId',
      )
    }
    return new SeamHttpConnectWebviews(constructorOptions)
  }

  async updateClientSessionToken(
    clientSessionToken: SeamHttpOptionsWithClientSessionToken['clientSessionToken'],
  ): Promise<void> {
    const { headers } = this.client.defaults
    const authHeaders = getAuthHeadersForClientSessionToken({
      clientSessionToken,
    })
    for (const key of Object.keys(authHeaders)) {
      if (headers[key] == null) {
        throw new Error(
          'Cannot update a clientSessionToken on a client created without a clientSessionToken',
        )
      }
    }
    this.client.defaults.headers = { ...headers, ...authHeaders }
    const clientSessions = SeamHttpClientSessions.fromClient(this.client)
    await clientSessions.get()
  }

  create(
    body?: ConnectWebviewsCreateBody,
  ): SeamHttpRequest<ConnectWebviewsCreateResponse, 'connect_webview'> {
    return new SeamHttpRequest(this, {
      path: '/connect_webviews/create',
      method: 'post',
      body,
      responseKey: 'connect_webview',
    })
  }

  delete(body?: ConnectWebviewsDeleteBody): SeamHttpRequest<void, undefined> {
    return new SeamHttpRequest(this, {
      path: '/connect_webviews/delete',
      method: 'post',
      body,
      responseKey: undefined,
    })
  }

  get(
    body?: ConnectWebviewsGetParams,
  ): SeamHttpRequest<ConnectWebviewsGetResponse, 'connect_webview'> {
    return new SeamHttpRequest(this, {
      path: '/connect_webviews/get',
      method: 'post',
      body,
      responseKey: 'connect_webview',
    })
  }

  list(
    body?: ConnectWebviewsListParams,
  ): SeamHttpRequest<ConnectWebviewsListResponse, 'connect_webviews'> {
    return new SeamHttpRequest(this, {
      path: '/connect_webviews/list',
      method: 'post',
      body,
      responseKey: 'connect_webviews',
    })
  }

  view(params?: ConnectWebviewsViewParams): SeamHttpRequest<void, undefined> {
    return new SeamHttpRequest(this, {
      path: '/connect_webviews/view',
      method: 'get',
      params,
      responseKey: undefined,
    })
  }
}

export type ConnectWebviewsCreateBody =
  RouteRequestBody<'/connect_webviews/create'>

export type ConnectWebviewsCreateResponse = SetNonNullable<
  Required<RouteResponse<'/connect_webviews/create'>>
>

export type ConnectWebviewsCreateOptions = never

export type ConnectWebviewsDeleteBody =
  RouteRequestBody<'/connect_webviews/delete'>

export type ConnectWebviewsDeleteResponse = SetNonNullable<
  Required<RouteResponse<'/connect_webviews/delete'>>
>

export type ConnectWebviewsDeleteOptions = never

export type ConnectWebviewsGetParams = RouteRequestBody<'/connect_webviews/get'>

export type ConnectWebviewsGetResponse = SetNonNullable<
  Required<RouteResponse<'/connect_webviews/get'>>
>

export type ConnectWebviewsGetOptions = never

export type ConnectWebviewsListParams =
  RouteRequestBody<'/connect_webviews/list'>

export type ConnectWebviewsListResponse = SetNonNullable<
  Required<RouteResponse<'/connect_webviews/list'>>
>

export type ConnectWebviewsListOptions = never

export type ConnectWebviewsViewParams =
  RouteRequestParams<'/connect_webviews/view'>

export type ConnectWebviewsViewResponse = SetNonNullable<
  Required<RouteResponse<'/connect_webviews/view'>>
>

export type ConnectWebviewsViewOptions = never
