// @seamapi/http/src/lib/seam/connect/routes/noise-sensors-noise-thresholds.ts
/*
 * Automatically generated by generate-routes.ts.
 * Do not edit this file or add other files to this directory.
 */

import type { RouteRequestBody, RouteResponse } from '@seamapi/types/connect'
import type { SetNonNullable } from 'type-fest'

import {
  getAuthHeadersForClientSessionToken,
  warnOnInsecureuserIdentifierKey,
} from 'lib/seam/connect/auth.js'
import { type Client, createClient } from 'lib/seam/connect/client.js'
import {
  isSeamHttpOptionsWithApiKey,
  isSeamHttpOptionsWithClient,
  isSeamHttpOptionsWithClientSessionToken,
  isSeamHttpOptionsWithConsoleSessionToken,
  isSeamHttpOptionsWithPersonalAccessToken,
  type SeamHttpFromPublishableKeyOptions,
  SeamHttpInvalidOptionsError,
  type SeamHttpOptions,
  type SeamHttpOptionsWithApiKey,
  type SeamHttpOptionsWithClient,
  type SeamHttpOptionsWithClientSessionToken,
  type SeamHttpOptionsWithConsoleSessionToken,
  type SeamHttpOptionsWithPersonalAccessToken,
  type SeamHttpRequestOptions,
} from 'lib/seam/connect/options.js'
import {
  limitToSeamHttpRequestOptions,
  parseOptions,
} from 'lib/seam/connect/parse-options.js'
import { SeamHttpRequest } from 'lib/seam/connect/seam-http-request.js'

import { SeamHttpClientSessions } from './client-sessions.js'

export class SeamHttpNoiseSensorsNoiseThresholds {
  client: Client
  readonly defaults: Required<SeamHttpRequestOptions>

  constructor(apiKeyOrOptions: string | SeamHttpOptions = {}) {
    const options = parseOptions(apiKeyOrOptions)
    this.client = 'client' in options ? options.client : createClient(options)
    this.defaults = limitToSeamHttpRequestOptions(options)
  }

  static fromClient(
    client: SeamHttpOptionsWithClient['client'],
    options: Omit<SeamHttpOptionsWithClient, 'client'> = {},
  ): SeamHttpNoiseSensorsNoiseThresholds {
    const constructorOptions = { ...options, client }
    if (!isSeamHttpOptionsWithClient(constructorOptions)) {
      throw new SeamHttpInvalidOptionsError('Missing client')
    }
    return new SeamHttpNoiseSensorsNoiseThresholds(constructorOptions)
  }

  static fromApiKey(
    apiKey: SeamHttpOptionsWithApiKey['apiKey'],
    options: Omit<SeamHttpOptionsWithApiKey, 'apiKey'> = {},
  ): SeamHttpNoiseSensorsNoiseThresholds {
    const constructorOptions = { ...options, apiKey }
    if (!isSeamHttpOptionsWithApiKey(constructorOptions)) {
      throw new SeamHttpInvalidOptionsError('Missing apiKey')
    }
    return new SeamHttpNoiseSensorsNoiseThresholds(constructorOptions)
  }

  static fromClientSessionToken(
    clientSessionToken: SeamHttpOptionsWithClientSessionToken['clientSessionToken'],
    options: Omit<
      SeamHttpOptionsWithClientSessionToken,
      'clientSessionToken'
    > = {},
  ): SeamHttpNoiseSensorsNoiseThresholds {
    const constructorOptions = { ...options, clientSessionToken }
    if (!isSeamHttpOptionsWithClientSessionToken(constructorOptions)) {
      throw new SeamHttpInvalidOptionsError('Missing clientSessionToken')
    }
    return new SeamHttpNoiseSensorsNoiseThresholds(constructorOptions)
  }

  static async fromPublishableKey(
    publishableKey: string,
    userIdentifierKey: string,
    options: SeamHttpFromPublishableKeyOptions = {},
  ): Promise<SeamHttpNoiseSensorsNoiseThresholds> {
    warnOnInsecureuserIdentifierKey(userIdentifierKey)
    const clientOptions = parseOptions({ ...options, publishableKey })
    if (isSeamHttpOptionsWithClient(clientOptions)) {
      throw new SeamHttpInvalidOptionsError(
        'The client option cannot be used with SeamHttp.fromPublishableKey',
      )
    }
    const client = createClient(clientOptions)
    const clientSessions = SeamHttpClientSessions.fromClient(client)
    const { token } = await clientSessions.getOrCreate({
      user_identifier_key: userIdentifierKey,
    })
    return SeamHttpNoiseSensorsNoiseThresholds.fromClientSessionToken(
      token,
      options,
    )
  }

  static fromConsoleSessionToken(
    consoleSessionToken: SeamHttpOptionsWithConsoleSessionToken['consoleSessionToken'],
    workspaceId: SeamHttpOptionsWithConsoleSessionToken['workspaceId'],
    options: Omit<
      SeamHttpOptionsWithConsoleSessionToken,
      'consoleSessionToken' | 'workspaceId'
    > = {},
  ): SeamHttpNoiseSensorsNoiseThresholds {
    const constructorOptions = { ...options, consoleSessionToken, workspaceId }
    if (!isSeamHttpOptionsWithConsoleSessionToken(constructorOptions)) {
      throw new SeamHttpInvalidOptionsError(
        'Missing consoleSessionToken or workspaceId',
      )
    }
    return new SeamHttpNoiseSensorsNoiseThresholds(constructorOptions)
  }

  static fromPersonalAccessToken(
    personalAccessToken: SeamHttpOptionsWithPersonalAccessToken['personalAccessToken'],
    workspaceId: SeamHttpOptionsWithPersonalAccessToken['workspaceId'],
    options: Omit<
      SeamHttpOptionsWithPersonalAccessToken,
      'personalAccessToken' | 'workspaceId'
    > = {},
  ): SeamHttpNoiseSensorsNoiseThresholds {
    const constructorOptions = { ...options, personalAccessToken, workspaceId }
    if (!isSeamHttpOptionsWithPersonalAccessToken(constructorOptions)) {
      throw new SeamHttpInvalidOptionsError(
        'Missing personalAccessToken or workspaceId',
      )
    }
    return new SeamHttpNoiseSensorsNoiseThresholds(constructorOptions)
  }

  async updateClientSessionToken(
    clientSessionToken: SeamHttpOptionsWithClientSessionToken['clientSessionToken'],
  ): Promise<void> {
    const { headers } = this.client.defaults
    const authHeaders = getAuthHeadersForClientSessionToken({
      clientSessionToken,
    })
    for (const key of Object.keys(authHeaders)) {
      if (headers[key] == null) {
        throw new Error(
          'Cannot update a clientSessionToken on a client created without a clientSessionToken',
        )
      }
    }
    this.client.defaults.headers = { ...headers, ...authHeaders }
    const clientSessions = SeamHttpClientSessions.fromClient(this.client)
    await clientSessions.get()
  }

  create(
    body?: NoiseSensorsNoiseThresholdsCreateBody,
  ): SeamHttpRequest<
    NoiseSensorsNoiseThresholdsCreateResponse,
    'noise_threshold'
  > {
    return new SeamHttpRequest(this, {
      path: '/noise_sensors/noise_thresholds/create',
      method: 'post',
      body,
      responseKey: 'noise_threshold',
    })
  }

  delete(
    body?: NoiseSensorsNoiseThresholdsDeleteBody,
  ): SeamHttpRequest<void, undefined> {
    return new SeamHttpRequest(this, {
      path: '/noise_sensors/noise_thresholds/delete',
      method: 'post',
      body,
      responseKey: undefined,
    })
  }

  get(
    body?: NoiseSensorsNoiseThresholdsGetParams,
  ): SeamHttpRequest<
    NoiseSensorsNoiseThresholdsGetResponse,
    'noise_threshold'
  > {
    return new SeamHttpRequest(this, {
      path: '/noise_sensors/noise_thresholds/get',
      method: 'post',
      body,
      responseKey: 'noise_threshold',
    })
  }

  list(
    body?: NoiseSensorsNoiseThresholdsListParams,
  ): SeamHttpRequest<
    NoiseSensorsNoiseThresholdsListResponse,
    'noise_thresholds'
  > {
    return new SeamHttpRequest(this, {
      path: '/noise_sensors/noise_thresholds/list',
      method: 'post',
      body,
      responseKey: 'noise_thresholds',
    })
  }

  update(
    body?: NoiseSensorsNoiseThresholdsUpdateBody,
  ): SeamHttpRequest<void, undefined> {
    return new SeamHttpRequest(this, {
      path: '/noise_sensors/noise_thresholds/update',
      method: 'post',
      body,
      responseKey: undefined,
    })
  }
}

export type NoiseSensorsNoiseThresholdsCreateBody =
  RouteRequestBody<'/noise_sensors/noise_thresholds/create'>

export type NoiseSensorsNoiseThresholdsCreateResponse = SetNonNullable<
  Required<RouteResponse<'/noise_sensors/noise_thresholds/create'>>
>

export type NoiseSensorsNoiseThresholdsCreateOptions = never

export type NoiseSensorsNoiseThresholdsDeleteBody =
  RouteRequestBody<'/noise_sensors/noise_thresholds/delete'>

export type NoiseSensorsNoiseThresholdsDeleteResponse = SetNonNullable<
  Required<RouteResponse<'/noise_sensors/noise_thresholds/delete'>>
>

export type NoiseSensorsNoiseThresholdsDeleteOptions = never

export type NoiseSensorsNoiseThresholdsGetParams =
  RouteRequestBody<'/noise_sensors/noise_thresholds/get'>

export type NoiseSensorsNoiseThresholdsGetResponse = SetNonNullable<
  Required<RouteResponse<'/noise_sensors/noise_thresholds/get'>>
>

export type NoiseSensorsNoiseThresholdsGetOptions = never

export type NoiseSensorsNoiseThresholdsListParams =
  RouteRequestBody<'/noise_sensors/noise_thresholds/list'>

export type NoiseSensorsNoiseThresholdsListResponse = SetNonNullable<
  Required<RouteResponse<'/noise_sensors/noise_thresholds/list'>>
>

export type NoiseSensorsNoiseThresholdsListOptions = never

export type NoiseSensorsNoiseThresholdsUpdateBody =
  RouteRequestBody<'/noise_sensors/noise_thresholds/update'>

export type NoiseSensorsNoiseThresholdsUpdateResponse = SetNonNullable<
  Required<RouteResponse<'/noise_sensors/noise_thresholds/update'>>
>

export type NoiseSensorsNoiseThresholdsUpdateOptions = never
