// @seamapi/http/src/lib/seam/connect/routes/thermostats-climate-setting-schedules.ts
/*
 * Automatically generated by generate-routes.ts.
 * Do not edit this file or add other files to this directory.
 */

import type { RouteRequestBody, RouteResponse } from '@seamapi/types/connect'
import type { SetNonNullable } from 'type-fest'

import {
  getAuthHeadersForClientSessionToken,
  warnOnInsecureuserIdentifierKey,
} from 'lib/seam/connect/auth.js'
import { type Client, createClient } from 'lib/seam/connect/client.js'
import {
  isSeamHttpOptionsWithApiKey,
  isSeamHttpOptionsWithClient,
  isSeamHttpOptionsWithClientSessionToken,
  isSeamHttpOptionsWithConsoleSessionToken,
  isSeamHttpOptionsWithPersonalAccessToken,
  type SeamHttpFromPublishableKeyOptions,
  SeamHttpInvalidOptionsError,
  type SeamHttpOptions,
  type SeamHttpOptionsWithApiKey,
  type SeamHttpOptionsWithClient,
  type SeamHttpOptionsWithClientSessionToken,
  type SeamHttpOptionsWithConsoleSessionToken,
  type SeamHttpOptionsWithPersonalAccessToken,
  type SeamHttpRequestOptions,
} from 'lib/seam/connect/options.js'
import {
  limitToSeamHttpRequestOptions,
  parseOptions,
} from 'lib/seam/connect/parse-options.js'
import { SeamHttpRequest } from 'lib/seam/connect/seam-http-request.js'

import { SeamHttpClientSessions } from './client-sessions.js'

export class SeamHttpThermostatsClimateSettingSchedules {
  client: Client
  readonly defaults: Required<SeamHttpRequestOptions>

  constructor(apiKeyOrOptions: string | SeamHttpOptions = {}) {
    const options = parseOptions(apiKeyOrOptions)
    this.client = 'client' in options ? options.client : createClient(options)
    this.defaults = limitToSeamHttpRequestOptions(options)
  }

  static fromClient(
    client: SeamHttpOptionsWithClient['client'],
    options: Omit<SeamHttpOptionsWithClient, 'client'> = {},
  ): SeamHttpThermostatsClimateSettingSchedules {
    const constructorOptions = { ...options, client }
    if (!isSeamHttpOptionsWithClient(constructorOptions)) {
      throw new SeamHttpInvalidOptionsError('Missing client')
    }
    return new SeamHttpThermostatsClimateSettingSchedules(constructorOptions)
  }

  static fromApiKey(
    apiKey: SeamHttpOptionsWithApiKey['apiKey'],
    options: Omit<SeamHttpOptionsWithApiKey, 'apiKey'> = {},
  ): SeamHttpThermostatsClimateSettingSchedules {
    const constructorOptions = { ...options, apiKey }
    if (!isSeamHttpOptionsWithApiKey(constructorOptions)) {
      throw new SeamHttpInvalidOptionsError('Missing apiKey')
    }
    return new SeamHttpThermostatsClimateSettingSchedules(constructorOptions)
  }

  static fromClientSessionToken(
    clientSessionToken: SeamHttpOptionsWithClientSessionToken['clientSessionToken'],
    options: Omit<
      SeamHttpOptionsWithClientSessionToken,
      'clientSessionToken'
    > = {},
  ): SeamHttpThermostatsClimateSettingSchedules {
    const constructorOptions = { ...options, clientSessionToken }
    if (!isSeamHttpOptionsWithClientSessionToken(constructorOptions)) {
      throw new SeamHttpInvalidOptionsError('Missing clientSessionToken')
    }
    return new SeamHttpThermostatsClimateSettingSchedules(constructorOptions)
  }

  static async fromPublishableKey(
    publishableKey: string,
    userIdentifierKey: string,
    options: SeamHttpFromPublishableKeyOptions = {},
  ): Promise<SeamHttpThermostatsClimateSettingSchedules> {
    warnOnInsecureuserIdentifierKey(userIdentifierKey)
    const clientOptions = parseOptions({ ...options, publishableKey })
    if (isSeamHttpOptionsWithClient(clientOptions)) {
      throw new SeamHttpInvalidOptionsError(
        'The client option cannot be used with SeamHttp.fromPublishableKey',
      )
    }
    const client = createClient(clientOptions)
    const clientSessions = SeamHttpClientSessions.fromClient(client)
    const { token } = await clientSessions.getOrCreate({
      user_identifier_key: userIdentifierKey,
    })
    return SeamHttpThermostatsClimateSettingSchedules.fromClientSessionToken(
      token,
      options,
    )
  }

  static fromConsoleSessionToken(
    consoleSessionToken: SeamHttpOptionsWithConsoleSessionToken['consoleSessionToken'],
    workspaceId: SeamHttpOptionsWithConsoleSessionToken['workspaceId'],
    options: Omit<
      SeamHttpOptionsWithConsoleSessionToken,
      'consoleSessionToken' | 'workspaceId'
    > = {},
  ): SeamHttpThermostatsClimateSettingSchedules {
    const constructorOptions = { ...options, consoleSessionToken, workspaceId }
    if (!isSeamHttpOptionsWithConsoleSessionToken(constructorOptions)) {
      throw new SeamHttpInvalidOptionsError(
        'Missing consoleSessionToken or workspaceId',
      )
    }
    return new SeamHttpThermostatsClimateSettingSchedules(constructorOptions)
  }

  static fromPersonalAccessToken(
    personalAccessToken: SeamHttpOptionsWithPersonalAccessToken['personalAccessToken'],
    workspaceId: SeamHttpOptionsWithPersonalAccessToken['workspaceId'],
    options: Omit<
      SeamHttpOptionsWithPersonalAccessToken,
      'personalAccessToken' | 'workspaceId'
    > = {},
  ): SeamHttpThermostatsClimateSettingSchedules {
    const constructorOptions = { ...options, personalAccessToken, workspaceId }
    if (!isSeamHttpOptionsWithPersonalAccessToken(constructorOptions)) {
      throw new SeamHttpInvalidOptionsError(
        'Missing personalAccessToken or workspaceId',
      )
    }
    return new SeamHttpThermostatsClimateSettingSchedules(constructorOptions)
  }

  async updateClientSessionToken(
    clientSessionToken: SeamHttpOptionsWithClientSessionToken['clientSessionToken'],
  ): Promise<void> {
    const { headers } = this.client.defaults
    const authHeaders = getAuthHeadersForClientSessionToken({
      clientSessionToken,
    })
    for (const key of Object.keys(authHeaders)) {
      if (headers[key] == null) {
        throw new Error(
          'Cannot update a clientSessionToken on a client created without a clientSessionToken',
        )
      }
    }
    this.client.defaults.headers = { ...headers, ...authHeaders }
    const clientSessions = SeamHttpClientSessions.fromClient(this.client)
    await clientSessions.get()
  }

  create(
    body?: ThermostatsClimateSettingSchedulesCreateBody,
  ): SeamHttpRequest<
    ThermostatsClimateSettingSchedulesCreateResponse,
    'climate_setting_schedule'
  > {
    return new SeamHttpRequest(this, {
      path: '/thermostats/climate_setting_schedules/create',
      method: 'post',
      body,
      responseKey: 'climate_setting_schedule',
    })
  }

  delete(
    body?: ThermostatsClimateSettingSchedulesDeleteBody,
  ): SeamHttpRequest<void, undefined> {
    return new SeamHttpRequest(this, {
      path: '/thermostats/climate_setting_schedules/delete',
      method: 'post',
      body,
      responseKey: undefined,
    })
  }

  get(
    body?: ThermostatsClimateSettingSchedulesGetParams,
  ): SeamHttpRequest<
    ThermostatsClimateSettingSchedulesGetResponse,
    'climate_setting_schedule'
  > {
    return new SeamHttpRequest(this, {
      path: '/thermostats/climate_setting_schedules/get',
      method: 'post',
      body,
      responseKey: 'climate_setting_schedule',
    })
  }

  list(
    body?: ThermostatsClimateSettingSchedulesListParams,
  ): SeamHttpRequest<
    ThermostatsClimateSettingSchedulesListResponse,
    'climate_setting_schedules'
  > {
    return new SeamHttpRequest(this, {
      path: '/thermostats/climate_setting_schedules/list',
      method: 'post',
      body,
      responseKey: 'climate_setting_schedules',
    })
  }

  update(
    body?: ThermostatsClimateSettingSchedulesUpdateBody,
  ): SeamHttpRequest<void, undefined> {
    return new SeamHttpRequest(this, {
      path: '/thermostats/climate_setting_schedules/update',
      method: 'post',
      body,
      responseKey: undefined,
    })
  }
}

export type ThermostatsClimateSettingSchedulesCreateBody =
  RouteRequestBody<'/thermostats/climate_setting_schedules/create'>

export type ThermostatsClimateSettingSchedulesCreateResponse = SetNonNullable<
  Required<RouteResponse<'/thermostats/climate_setting_schedules/create'>>
>

export type ThermostatsClimateSettingSchedulesCreateOptions = never

export type ThermostatsClimateSettingSchedulesDeleteBody =
  RouteRequestBody<'/thermostats/climate_setting_schedules/delete'>

export type ThermostatsClimateSettingSchedulesDeleteResponse = SetNonNullable<
  Required<RouteResponse<'/thermostats/climate_setting_schedules/delete'>>
>

export type ThermostatsClimateSettingSchedulesDeleteOptions = never

export type ThermostatsClimateSettingSchedulesGetParams =
  RouteRequestBody<'/thermostats/climate_setting_schedules/get'>

export type ThermostatsClimateSettingSchedulesGetResponse = SetNonNullable<
  Required<RouteResponse<'/thermostats/climate_setting_schedules/get'>>
>

export type ThermostatsClimateSettingSchedulesGetOptions = never

export type ThermostatsClimateSettingSchedulesListParams =
  RouteRequestBody<'/thermostats/climate_setting_schedules/list'>

export type ThermostatsClimateSettingSchedulesListResponse = SetNonNullable<
  Required<RouteResponse<'/thermostats/climate_setting_schedules/list'>>
>

export type ThermostatsClimateSettingSchedulesListOptions = never

export type ThermostatsClimateSettingSchedulesUpdateBody =
  RouteRequestBody<'/thermostats/climate_setting_schedules/update'>

export type ThermostatsClimateSettingSchedulesUpdateResponse = SetNonNullable<
  Required<RouteResponse<'/thermostats/climate_setting_schedules/update'>>
>

export type ThermostatsClimateSettingSchedulesUpdateOptions = never
