// @seamapi/http/lib/seam/connect/routes/user-identities.js
/*
 * Automatically generated by generate-routes.ts.
 * Do not edit this file or add other files to this directory.
 */
import { getAuthHeadersForClientSessionToken, warnOnInsecureuserIdentifierKey, } from '../../../../lib/seam/connect/auth.js';
import { createClient } from '../../../../lib/seam/connect/client.js';
import { isSeamHttpOptionsWithApiKey, isSeamHttpOptionsWithClient, isSeamHttpOptionsWithClientSessionToken, isSeamHttpOptionsWithConsoleSessionToken, isSeamHttpOptionsWithPersonalAccessToken, SeamHttpInvalidOptionsError, } from '../../../../lib/seam/connect/options.js';
import { limitToSeamHttpRequestOptions, parseOptions, } from '../../../../lib/seam/connect/parse-options.js';
import { SeamHttpRequest } from '../../../../lib/seam/connect/seam-http-request.js';
import { SeamHttpClientSessions } from './client-sessions.js';
import { SeamHttpUserIdentitiesEnrollmentAutomations } from './user-identities-enrollment-automations.js';
export class SeamHttpUserIdentities {
    constructor(apiKeyOrOptions = {}) {
        const options = parseOptions(apiKeyOrOptions);
        this.client = 'client' in options ? options.client : createClient(options);
        this.defaults = limitToSeamHttpRequestOptions(options);
    }
    static fromClient(client, options = {}) {
        const constructorOptions = { ...options, client };
        if (!isSeamHttpOptionsWithClient(constructorOptions)) {
            throw new SeamHttpInvalidOptionsError('Missing client');
        }
        return new SeamHttpUserIdentities(constructorOptions);
    }
    static fromApiKey(apiKey, options = {}) {
        const constructorOptions = { ...options, apiKey };
        if (!isSeamHttpOptionsWithApiKey(constructorOptions)) {
            throw new SeamHttpInvalidOptionsError('Missing apiKey');
        }
        return new SeamHttpUserIdentities(constructorOptions);
    }
    static fromClientSessionToken(clientSessionToken, options = {}) {
        const constructorOptions = { ...options, clientSessionToken };
        if (!isSeamHttpOptionsWithClientSessionToken(constructorOptions)) {
            throw new SeamHttpInvalidOptionsError('Missing clientSessionToken');
        }
        return new SeamHttpUserIdentities(constructorOptions);
    }
    static async fromPublishableKey(publishableKey, userIdentifierKey, options = {}) {
        warnOnInsecureuserIdentifierKey(userIdentifierKey);
        const clientOptions = parseOptions({ ...options, publishableKey });
        if (isSeamHttpOptionsWithClient(clientOptions)) {
            throw new SeamHttpInvalidOptionsError('The client option cannot be used with SeamHttp.fromPublishableKey');
        }
        const client = createClient(clientOptions);
        const clientSessions = SeamHttpClientSessions.fromClient(client);
        const { token } = await clientSessions.getOrCreate({
            user_identifier_key: userIdentifierKey,
        });
        return SeamHttpUserIdentities.fromClientSessionToken(token, options);
    }
    static fromConsoleSessionToken(consoleSessionToken, workspaceId, options = {}) {
        const constructorOptions = { ...options, consoleSessionToken, workspaceId };
        if (!isSeamHttpOptionsWithConsoleSessionToken(constructorOptions)) {
            throw new SeamHttpInvalidOptionsError('Missing consoleSessionToken or workspaceId');
        }
        return new SeamHttpUserIdentities(constructorOptions);
    }
    static fromPersonalAccessToken(personalAccessToken, workspaceId, options = {}) {
        const constructorOptions = { ...options, personalAccessToken, workspaceId };
        if (!isSeamHttpOptionsWithPersonalAccessToken(constructorOptions)) {
            throw new SeamHttpInvalidOptionsError('Missing personalAccessToken or workspaceId');
        }
        return new SeamHttpUserIdentities(constructorOptions);
    }
    async updateClientSessionToken(clientSessionToken) {
        const { headers } = this.client.defaults;
        const authHeaders = getAuthHeadersForClientSessionToken({
            clientSessionToken,
        });
        for (const key of Object.keys(authHeaders)) {
            if (headers[key] == null) {
                throw new Error('Cannot update a clientSessionToken on a client created without a clientSessionToken');
            }
        }
        this.client.defaults.headers = { ...headers, ...authHeaders };
        const clientSessions = SeamHttpClientSessions.fromClient(this.client);
        await clientSessions.get();
    }
    get enrollmentAutomations() {
        return SeamHttpUserIdentitiesEnrollmentAutomations.fromClient(this.client, this.defaults);
    }
    addAcsUser(body) {
        return new SeamHttpRequest(this, {
            path: '/user_identities/add_acs_user',
            method: 'post',
            body,
            responseKey: undefined,
        });
    }
    create(body) {
        return new SeamHttpRequest(this, {
            path: '/user_identities/create',
            method: 'post',
            body,
            responseKey: 'user_identity',
        });
    }
    delete(body) {
        return new SeamHttpRequest(this, {
            path: '/user_identities/delete',
            method: 'post',
            body,
            responseKey: undefined,
        });
    }
    get(body) {
        return new SeamHttpRequest(this, {
            path: '/user_identities/get',
            method: 'post',
            body,
            responseKey: 'user_identity',
        });
    }
    grantAccessToDevice(body) {
        return new SeamHttpRequest(this, {
            path: '/user_identities/grant_access_to_device',
            method: 'post',
            body,
            responseKey: undefined,
        });
    }
    list(body) {
        return new SeamHttpRequest(this, {
            path: '/user_identities/list',
            method: 'post',
            body,
            responseKey: 'user_identities',
        });
    }
    listAccessibleDevices(body) {
        return new SeamHttpRequest(this, {
            path: '/user_identities/list_accessible_devices',
            method: 'post',
            body,
            responseKey: 'accessible_devices',
        });
    }
    listAcsSystems(body) {
        return new SeamHttpRequest(this, {
            path: '/user_identities/list_acs_systems',
            method: 'post',
            body,
            responseKey: 'acs_systems',
        });
    }
    listAcsUsers(body) {
        return new SeamHttpRequest(this, {
            path: '/user_identities/list_acs_users',
            method: 'post',
            body,
            responseKey: 'acs_users',
        });
    }
    removeAcsUser(body) {
        return new SeamHttpRequest(this, {
            path: '/user_identities/remove_acs_user',
            method: 'post',
            body,
            responseKey: undefined,
        });
    }
    revokeAccessToDevice(body) {
        return new SeamHttpRequest(this, {
            path: '/user_identities/revoke_access_to_device',
            method: 'post',
            body,
            responseKey: undefined,
        });
    }
    update(body) {
        return new SeamHttpRequest(this, {
            path: '/user_identities/update',
            method: 'post',
            body,
            responseKey: undefined,
        });
    }
}
